File: Review-14-Python-Builder-Pattern/Example-3/room.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Flooring:
    type_name: str = "Generic"
    unit_cost: float = 1.0


@dataclass
class DimensionSet:
    length: float = 1.0
    width: float = 1.0


class Room:
    def __init__(self, nme: str = "Generic"):
        self.__name = nme
        self.__dimensions = DimensionSet()
        self.__flooring = Flooring()

    @property
    def name(self):
        """
        Get the name using a property.

        @property must come before @...setter...
        """

        return self.__name

    def set_flooring(self, nme: str, unit_c: float):
        """
        Set the flooring.

        Args:
            nme: flooring type name
            unit_c: unit cost

        """

        self.__flooring.type_name = nme
        self.__flooring.unit_cost = unit_c

    def area(self) -> float:
        """
        Compute the area of flooring for a room.
        """

        return float(self.__dimensions.width) * self.__dimensions.length

    def flooring_cost(self) -> float:
        """
        Compute the flooring cost based on `self.area()` and unit cost.
        """

        return self.area() * self.__flooring.unit_cost

    def __str__(self) -> str:
        return "\n".join(
            (
                f"Room ({self.name})",
                "  {:<6}: {:>8.1f}".format("Length", self.__dimensions.length),
                "  {:<6}: {:>8.1f}".format("Width", self.__dimensions.width),
                "  {:<6}: {:>8.1f}".format("Area", self.area()),
                "",
                f"  Flooring  : {self.__flooring.type_name}",
                f"  Unit Cost : $ {self.__flooring.unit_cost:>8.2f}",
                "  Total Cost: $ {:>8.2f}".format(self.flooring_cost()),
                "\n",
            )
        )

    def __lt__(self, rhs) -> bool:
        if self.name == rhs.name:
            return self.area() < rhs.area()  # pylint caught the missing return

        return self.name < rhs.name

    def __eq__(self, rhs) -> bool:
        if not isinstance(rhs, Room):
            return False

        return self.name == (rhs.name) and self.area() == rhs.area()


class RoomBuilder:
    def __init__(self):
        self.__name = None
        self.__length = None
        self.__width = None
        self.__flooring = None

    def with_name(self, nme: str) -> RoomBuilder:
        """
        Set the name using the builder pattern.

        Args:
            nme: room name
        """

        self.__name = nme.strip()

        return self

    def with_flooring(self, nme: str, unit_c: float) -> RoomBuilder:
        """
        Set the Flooring using the builder pattern.

        Args
            nme: flooring type name
            unit_c: unit cost

        """

        flr = Flooring(nme, unit_c)
        self.__flooring = flr

        return self

    def with_dimensions(self, l: float, w: float) -> RoomBuilder:
        """
        Set the Flooring using the builder pattern.

        Args:
            l: length
            w: width

        """

        self.__length = l
        self.__width = w

        return self

    def validate_name(self) -> None:
        """
        Raise a Value Error if:
          1. Name was not set
          2. Name is the empty string
          3. Name has fewer than 3 characters
        """

        if not self.__name:
            raise ValueError("No name was set")

        if len(self.__name) < 3:
            raise ValueError("Name len('{self.__name}') < 3")

    def build(self) -> Room:
        self.validate_name()

        if not self.__flooring:
            raise ValueError("No flooring was set")

        if not self.__length:
            raise ValueError("No length was set")

        if not self.__width:
            raise ValueError("No width was set")

        room = Room(self.__name)
        room._Room__flooring = self.__flooring
        room._Room__dimensions = DimensionSet(self.__length, self.__width)

        return room

File: 14-Rust-Builder-Pattern/Example-4/src/room.rs
use crate::error::*;
use crate::flooring::*;
use std::cmp::Ordering;
use std::fmt;
use std::fmt::Display; //,Formatter,Result};

#[derive(Clone, Debug)]
pub struct DimensionSet {
    pub length: f64,
    pub width: f64,
}

impl DimensionSet {
    pub fn new(l: f64, w: f64) -> Self {
        DimensionSet {
            length: l,
            width: w,
        }
    }
}

impl Default for DimensionSet {
    fn default() -> Self {
        DimensionSet::new(1f64, 1f64)
    }
}

impl From<(f64, f64)> for DimensionSet {
    fn from(dims: (f64, f64)) -> Self {
        DimensionSet::new(dims.0, dims.1)
    }
}

#[derive(Clone, Debug)]
pub struct Room {
    pub name: String,
    pub dimensions: DimensionSet,
    pub flooring: Flooring,
}

impl Room {
    pub fn builder() -> RoomBuilder<NoName, NoDimensions, NoFlooring> {
        RoomBuilder::new()
    }

    /// Compute the area of flooring for a room.
    pub fn area(&self) -> f64 {
        self.dimensions.width * self.dimensions.length
    }

    /// Compute the flooring cost based on `self.area()` and unit cost.
    pub fn flooring_cost(&self) -> f64 {
        self.area() * self.flooring.unit_cost
    }
}

impl Default for Room {
    fn default() -> Self {
        Room::builder()
            .with_name("Generic")
            .with_dimensions(1.0, 1.0)
            .unwrap()
            .with_flooring(Flooring::default())
            .build()
    }
}

impl Display for Room {
    #[allow(unused_must_use)]
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Room ({})", self.name);
        writeln!(f, "  {:<6}: {:>8.1}", "Length", self.dimensions.length);
        writeln!(f, "  {:<6}: {:>8.1}", "Width", self.dimensions.width);
        writeln!(f, "  {:<6}: {:>8.1}", "Area", self.area());
        writeln!(f);
        writeln!(f, "  Flooring  : {}", self.flooring.type_name);
        writeln!(f, "  Unit Cost : $ {:>8.2}", self.flooring.unit_cost);
        writeln!(f, "  Total Cost: $ {:>8.2}", self.flooring_cost())
    }
}

impl PartialOrd for Room {
    fn partial_cmp(&self, rhs: &Self) -> Option<Ordering> {
        if self.name.eq(&rhs.name) {
            self.area().partial_cmp(&rhs.area())
        } else {
            self.name.partial_cmp(&rhs.name)
        }
    }
}

impl PartialEq for Room {
    fn eq(&self, rhs: &Self) -> bool {
        self.name.eq(&rhs.name) && self.area().eq(&rhs.area())
    }
}

//------------------------------------------------------------------------------
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct NoName;

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct NoDimensions;

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct NoFlooring;

/// The RoomBuilder's state is defined by three generics
///   - SN - (State Name) whether the name is set
///   - SD - (State Dimensions) whether the length and width are set
///   - SF - (State Flooring) whether the flooring is set
///
#[derive(Clone, Debug, Default, PartialEq)]
pub struct RoomBuilder<SN, SD, SF> {
    name: SN,
    length: SD,
    width: SD,
    flooring: SF,
}

impl RoomBuilder<NoName, NoDimensions, NoFlooring> {
    pub fn new() -> Self {
        RoomBuilder {
            name: NoName,
            length: NoDimensions,
            width: NoDimensions,
            flooring: NoFlooring,
        }
    }
}

impl RoomBuilder<NoName, NoDimensions, NoFlooring> {
    pub fn from_existing(self, room: &Room) -> RoomBuilder<String, f64, Flooring> {
        RoomBuilder {
            name: room.name.clone(),
            width: room.dimensions.width,
            length: room.dimensions.length,
            flooring: room.flooring.clone(),
        }
    }

    pub fn with_name(self, nme: &str) -> RoomBuilder<String, NoDimensions, NoFlooring> {
        RoomBuilder {
            name: nme.into(),
            length: self.length,
            width: self.width,
            flooring: self.flooring,
        }
    }
}

impl RoomBuilder<String, NoDimensions, NoFlooring> {
    pub fn with_dimensions(
        self,
        length: f64,
        width: f64,
    ) -> Result<RoomBuilder<String, f64, NoFlooring>, BuildErrorWithState<Self>> {
        match (length > 0.0, width > 0.0) {
            (false, false) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'length' and 'width' must be > 0"),
                the_builder: self,
            }),
            (false, true) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'length' must be > 0"),
                the_builder: self,
            }),
            (true, false) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'width' must be > 0"),
                the_builder: self,
            }),
            (true, true) => Ok(RoomBuilder {
                name: self.name,
                length: length,
                width: width,
                flooring: self.flooring,
            }),
        }
    }
}

impl<SF> RoomBuilder<String, f64, SF> {
    pub fn with_flooring(self, flooring: Flooring) -> RoomBuilder<String, f64, Flooring> {
        RoomBuilder {
            name: self.name,
            length: self.length,
            width: self.width,
            flooring: flooring,
        }
    }
}

impl RoomBuilder<String, f64, Flooring> {
    pub fn with_dimensions(
        mut self,
        length: f64,
        width: f64,
    ) -> Result<Self, BuildErrorWithState<Self>> {
        match (length > 0.0, width > 0.0) {
            (false, false) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'length' and 'width' must be > 0"),
                the_builder: self,
            }),
            (false, true) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'length' must be > 0"),
                the_builder: self,
            }),
            (true, false) => Err(BuildErrorWithState {
                the_error: BuildError::InvalidDim("'width' must be > 0"),
                the_builder: self,
            }),
            (true, true) => {
                self.length = length;
                self.width = width;

                Ok(self)
            }
        }
    }

    pub fn build(self) -> Room {
        Room {
            name: self.name,
            dimensions: DimensionSet::new(self.length, self.width),
            flooring: self.flooring,
        }
    }
}
